Firestore/core/src/firebase/firestore/immutable/llrb_node_iterator.h

```diff
@@ -31,12 +31,35 @@ namespace immutable {
 namespace impl {
 
 /**
- * An iterator for traversing LlrbNodes.
+ * A forward iterator for traversing LlrbNodes. LlrbNodes represent the nodes
+ * in a tree implementing a sorted map so iterating with LlrbNodeIterator is
+ * an in-order traversal of the map.
  *
- * LlrbNode is an immutable tree, where insertions create new trees without
+ * ## Complexity
+ *
+ * LlrbNode is an immutable tree, where mutations create new trees without
  * invalidating any of the old instances. This means the tree cannot contain
  * parent pointers and thus this iterator implementation must keep an explicit
  * stack.
+ *
+ * For an underlying tree of size `n`:
+ *
+ *   * LlrbNodeIterator uses `O(lg(n))` memory for its stack, and
+ *   * incrementing an iterator is an `O(lg(n))` operation.
+ *
+ * ## Invalidation and Comparison
+ *
+ * LlrbNodeIterators compare based on the identity of the nodes themselves,
+ * not based on the values of the keys in the nodes. When adding and removing
+ * the same key and iterator obtained before and after will not compare equal.
+ *
+ * LlrbNodeIterators are not invalidated in any conventional sense because
+ * mutations of the underlying tree create new trees. Together this means that
+ * any given version of the tree can be iterated over from the same iterator
+ * repeatedly, but a "mutable view" of the tree kept by replacing the pointer
+ * to the root is effectively invalidated on each mutation.
+ *
+ * Note: LlrbNodeIterator does not extend the lifetime of its underlying tree.
```

Comment:
Hmm, I might be misreading it, but it appears that once a node is erased, `LlrbNodeIterator` does become invalidated, since it stores plain pointers to nodes, not shared pointers. I'd suggest adding that piece to the section on invalidation (potentially removing it from here). I'm fine if you do this in a later PR that has `erase` implementation. Feel free to push back if I'm wrong on this.

Reply:
`erase` also makes a new tree that just does not include the erased node, but the existing tree is unchanged. An iterator pointing at a node that has been "erased" will continue to be valid so long as the old version of the tree is kept alive.

This concern is not new; it exists in the Objective-C implementation already and leads to code like this:

https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/Local/FSTLocalDocumentsView.mm#L133